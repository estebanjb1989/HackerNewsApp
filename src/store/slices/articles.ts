import { createSlice, createAsyncThunk, current } from "@reduxjs/toolkit";
import notifee from '@notifee/react-native';
import { IArticle, IState } from "@interfaces";

interface IArticleState {
  articles: IArticle[];
  deletedArticles: string[];
  favoriteArticles: string[];
  loading: boolean;
  query: string;
}

const initialState = {
  articles: [],
  deletedArticles: [],
  favoriteArticles: [],
  loading: false,
  query: "tech"
} as IArticleState;

const articlesSlice = createSlice({
  name: "articlesSlice",
  initialState,
  reducers: {
    setLoading(state, action) {
      state.loading = action.payload
    },
    setArticles(state, action) {
      state.articles = action.payload;
    },
    deleteArticle(state, action) {
      state.deletedArticles = [...(state.deletedArticles || []), action.payload]
    },
    addToFavorites(state, action) {
      state.favoriteArticles = [...(state.favoriteArticles || []), action.payload]
    },
    removeFromFavorites(state, action) {
      state.favoriteArticles = state.favoriteArticles?.filter(fa => fa !== action.payload)
    },
    setQuery(state, action) {
      state.query = action.payload
    }
  },
});

// Define an async action creator
export const fetchArticles = createAsyncThunk(
  'articles/fetch',
  async (args, { dispatch, getState }) => {
    try {
      dispatch(setLoading(true))
      const query = (getState() as IState).articles?.query

      const response = await fetch("https://hn.algolia.com/api/v1/search_by_date?query=" + query)
      const data = await response.json()

      const articles = data.hits
        .filter((hit: any) => hit.story_title && hit.story_url)
        .map((hit: any) => ({
          id: hit.created_at_i,
          title: hit.story_title,
          author: hit.author,
          createdAt: hit.created_at,
          url: hit.story_url
        } as IArticle))
        .sort((a: any, b: any) => {
          if (a.created_at_i > b.created_at_i) return -1;
          if (a.created_at_i < b.created_at_i) return 1;
          return 0;
        })

      const notificationsEnabled = (getState() as IState)?.settings?.notificationsEnabled
      const currentArticles = (getState() as IState)?.articles?.articles

      if (notificationsEnabled && currentArticles?.length > 0) {
        const newArticles = articles.filter((art: IArticle) =>
          !currentArticles?.some((curr: IArticle) => curr.id === art.id)
        )

        const channelId = await notifee.createChannel({
          id: 'default',
          name: 'Default Channel',
        });

        for (const newArticle of newArticles) {
          await notifee.displayNotification({
            title: "New article by " + newArticle.author,
            body: newArticle.title,
            data: {
              url: newArticle.url || ""
            },
            android: {
              channelId,
              smallIcon: 'ic_launcher', // optional, defaults to 'ic_launcher'.
              // pressAction is needed if you want the notification to open the app when pressed
              pressAction: {
                id: 'default',
              },
            },
          });
        }
      }

      return articles
    } catch (err) {
      console.error(err)
      throw err
    } finally {
      dispatch(setLoading(false))
    }
  }
);

export const {
  setLoading,
  setArticles,
  deleteArticle,
  addToFavorites,
  removeFromFavorites,
  setQuery
} = articlesSlice.actions;

export default articlesSlice.reducer;
