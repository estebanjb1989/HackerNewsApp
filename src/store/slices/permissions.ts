import { createSlice } from "@reduxjs/toolkit";

interface IPermissionsState {
  notificationsPermissionStatus: number | null;
}

const initialState = {
  notificationsPermissionStatus: null,
} as IPermissionsState;

const permissionsSlice = createSlice({
  name: "permissionsSlice",
  initialState,
  reducers: {
    setNotificationsPermissionStatus(state, action) {
      state.notificationsPermissionStatus = action.payload
    },    
  },
});

export const {
  setNotificationsPermissionStatus
} = permissionsSlice.actions;

export default permissionsSlice.reducer;
