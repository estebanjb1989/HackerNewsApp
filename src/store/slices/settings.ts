import { createSlice } from "@reduxjs/toolkit";

interface ISettingsState {
  notificationsEnabled: boolean;
}

const initialState = {
  notificationsEnabled: false,
} as ISettingsState;

const settingsSlice = createSlice({
  name: "settingsSlice",
  initialState,
  reducers: {
    setNotificationsEnabled(state, action) {
      state.notificationsEnabled = action.payload
    },    
  },
});

export const {
  setNotificationsEnabled
} = settingsSlice.actions;

export default settingsSlice.reducer;
