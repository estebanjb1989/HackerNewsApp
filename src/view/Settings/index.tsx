import React from 'react';
import {
    SafeAreaView,
    StyleSheet,
    Switch,
    View,
    Text,
} from 'react-native';
import Title from '@component/Title';
import { useDispatch, useSelector } from 'react-redux';
import { setNotificationsEnabled } from '@store/slices/settings';

function Settings() {
    const dispatch = useDispatch();
    const notificationsEnabled = useSelector(state => state.settings.notificationsEnabled);

    const handleToggle = () => {
        dispatch(setNotificationsEnabled(!notificationsEnabled));
    };

    return (
        <SafeAreaView>
            <Title text="Settings" />
            <View style={styles.optionContainer}>
                <Switch
                    testID="settings_notification-enabled-toggle"
                    value={notificationsEnabled}
                    onValueChange={handleToggle}
                    trackColor={{ false: '#ccc', true: '#4CAF50' }}
                    thumbColor={notificationsEnabled ? '#fff' : '#fff'}
                />
                <Text style={styles.optionText}>Notifications Enabled</Text>
            </View>

        </SafeAreaView>
    );
}

const styles = StyleSheet.create({
    optionContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingHorizontal: 12,
        marginTop: 8,
    },
    optionText: {
        width: 240,
        marginLeft: 12,
    },
});

export default Settings;
