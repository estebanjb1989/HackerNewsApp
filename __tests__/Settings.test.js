import React from 'react';
import { render, fireEvent } from '@testing-library/react-native'
import Settings from "@view/Settings"

import * as settingsSlice from "@store/slices/settings"

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

const spySetNotificationsEnabled = jest.spyOn(settingsSlice,
  "setNotificationsEnabled"
)

jest.mock("@store/slices/settings", () => ({
  setNotificationsEnabled: jest.fn(),
}))

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn()
}))

describe("Settings view", () => {  
  it('renders correctly', () => {
    const tree = render(<Settings />);
    expect(tree).toMatchSnapshot();
  });

  it('should call set notifications enabled if toggle is set to true', async () => {
    const tree = render(<Settings />);
    const toggle = await tree.findByTestId("settings_notification-enabled-toggle")
    fireEvent(toggle, "onValueChange", true)
    expect(spySetNotificationsEnabled).toHaveBeenCalled()
  });

})

