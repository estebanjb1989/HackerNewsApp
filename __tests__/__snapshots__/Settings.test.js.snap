// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings view renders correctly 1`] = `
<RCTSafeAreaView>
  <Text
    style={
      [
        {
          "fontSize": 20,
          "marginHorizontal": 8,
          "marginVertical": 8,
          "minWidth": 240,
        },
        undefined,
      ]
    }
  >
    Settings
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingHorizontal": 4,
      }
    }
  >
    <RCTSwitch
      accessibilityRole="switch"
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onTintColor="#4CAF50"
      style={
        {
          "height": 31,
          "width": 51,
        }
      }
      testID="settings_notification-enabled-toggle"
      thumbTintColor="#fff"
      tintColor="#ccc"
      value={false}
    />
    <Text
      style={
        {
          "width": 240,
        }
      }
    >
      Notifications Enabled
    </Text>
  </View>
</RCTSafeAreaView>
`;
